kentik_api_library/kentik_api/analytics/data_frame_cache.py

```diff
@@ -0,0 +1,256 @@
+        log.debug("oldest: %s newest: %s", self.oldest, self.newest)
+
+    def __repr__(self) -> str:
+        return f"DFCache: dir: {self.data_dir.resolve()}"
+
+    @property
+    def files(self) -> Generator[Path, None, None]:
+        for f in sorted(self.data_dir.glob(self.filename_format.format(start="*", end="*"))):
+            yield f
+
+    @property
+    def file_count(self) -> int:
+        return len(list(self.files))
+
+    @property
+    def data_size(self) -> int:
+        return sum(f.stat().st_size for f in self.files)
+
+    @property
+    def is_empty(self) -> bool:
+        return self.file_count == 0
+
+    @property
+    def oldest(self) -> Optional[datetime]:
+        match = self.filename_format.format(start="*", end="*")
+        try:
+            first = sorted(list(self.data_dir.glob(match)))[0]
+        except IndexError:
+            log.debug("No match for %s in data directory %s", match, self.data_dir)
+            return None
+        start, end = self.parse_df_filename(first)
+        if start is None or end is None:
+            log.critical("Invalid filename %s in data directory %s", first, self.data_dir)
+            return None
+        return start
+
+    @property
+    def newest(self) -> Optional[datetime]:
+        match = self.filename_format.format(start="*", end="*")
+        try:
+            first = sorted(list(self.data_dir.glob(match)))[-1]
+        except IndexError:
+            log.debug("No match for %s in data directory %s", match, self.data_dir)
+            return None
+        start, end = self.parse_df_filename(first)
+        if start is None or end is None:
+            log.critical("Invalid filename %s in data directory %s", first, self.data_dir)
+            return None
+        return end
+
+    def info(self, out=None):
+        if out is None:
+            out = sys.stdout
+        print(f"files: {self.file_count}, total size: {self.data_size / 1000000} MB")
+        print(f"oldest data: {self.oldest}", file=out)
+        print(f"newest data: {self.newest}", file=out)
+
+    def files_in_range(
+        self, start: Optional[datetime] = None, end: Optional[datetime] = None, contained: bool = False
+    ) -> Generator[Path, None, None]:
+        if start is not None and end is not None and not start < end:
+            raise RuntimeError(f"Invalid arguments start {start} >= end {end}")
+        if self.is_empty:
+            log.debug("files_in_range: No files data directory: %s", self.data_dir)
+            return
+        if start is None:
+            start = self.oldest
+        if end is None:
+            end = self.newest
+        for f in self.files:
+            fs, fe = self.parse_df_filename(f)
+            if fs is None or fe is None:
+                continue
+            if contained:
+                if fs >= start and fe <= end:  # type: ignore
+                    yield f
+            elif fs <= end and fe > start:  # type: ignore
```

Comment:
I don't quite understand the case here

Reply:
Well, there are 2 cases:
- in the `contained == True` case only files with start and end times fully within the specified interval are returned
- otherwise all files with start and end time (even partially) overlapping the requested interval are returned.

Reply:
OK so this is about containing vs overlapping datetime ranges. Can you please add relevant comment?

Reply:
Will do.